models/articles.models.js

```diff
@@ -1,24 +1,58 @@
 const db = require("../db/connection");
-const { checkUserExists } = require("../utils/utils.js");
+const { checkItemExists } = require("../utils/utils.js");
 
-exports.fetchArticles = () => {
-  const queryString = `
+exports.fetchArticles = (sort_by, order, topic) => {
+  const validColumns = [
+    "article_id",
+    "title",
+    "topic",
+    "author",
+    "body",
+    "created_at",
+    "votes",
+  ];
+
+  if (!sort_by) sort_by = "created_at";
+  if (!order) order = "desc";
+
+  let queryString = `
     SELECT articles.article_id, articles.author, articles.title, articles.topic, articles.created_at, articles.votes, CAST(COUNT(comments.comment_id) AS INT) AS comment_count 
     FROM articles
-    LEFT JOIN comments ON articles.article_id = comments.article_id
-    GROUP BY articles.article_id
-    ORDER BY articles.created_at DESC;
-  `;
-  return db.query(queryString).then((result) => result.rows);
+    LEFT JOIN comments ON articles.article_id = comments.article_id`;
+  const queryStringOrderByGroupBy = ` GROUP BY articles.article_id ORDER BY ${sort_by} ${order};`;
+
+  return Promise.all([])
+    .then(() => {
+      if (topic !== undefined) {
+        return checkItemExists(topic, "topic", "topics", "slug");
+      }
+      if (!validColumns.includes(sort_by)) {
+        return Promise.reject({ status: 404, msg: "sort_by column not found"})
+      }
+      if (!["asc", "desc"].includes(order)) {
+        return Promise.reject({ status: 400, msg: "invalid order"})
+      }
+    })
+    .then(() => {
+      if (topic !== undefined) {
+        queryString += ` WHERE topic = '${topic}'`;
+      }
+      queryString += queryStringOrderByGroupBy;
+      return db.query(queryString);
+    })
+    .then((result) => {
+      return result.rows;
+    });
 };
 
 exports.fetchArticleById = (article_id) => {
   const queryString = `
-    SELECT users.username AS author, articles.title, articles.article_id, articles.topic, articles.body, articles.created_at, articles.votes
-    FROM articles
-    LEFT JOIN users ON articles.author = users.username
-    WHERE article_id = $1;
-  `;
+  SELECT articles.author, articles.title, articles.article_id, articles.topic, articles.body, articles.created_at, articles.votes, CAST(COUNT(comments.comment_id) AS INT) AS comment_count
+  FROM articles
+  LEFT JOIN comments ON articles.article_id = comments.article_id
+  WHERE articles.article_id = $1
+  GROUP BY articles.author, articles.title, articles.article_id, articles.topic, articles.body, articles.created_at, articles.votes;
```

Comment:
We actually only need to group by article.article_id here as that is the unique identifier.

Reply:
Done